pep-0611.rst

```diff
@@ -0,0 +1,257 @@
+PEP: 611
+Title: The one million limit
+Author: Mark Shannon <mark@hotpy.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 05-Dec-2019
+Post-History: 
+
+
+
+Abstract
+========
+This PR proposes a limit of one million (1 000 000) for various aspects of Python code and its implementation.
+
+The Python language does not specify limits for many of its features.
+Not having any limit to these values seems to enhance programmer freedom,
+at least superficially, but in practice the CPython VM and other Python virtual
+machines have implicit limits or are forced to assume that the limits are 
+astronomical, which is expensive.
+
+This PR lists a number of features which are to have a limit of one million.
+
+Motivation
+==========
+
+There are many values that need to be represented in a virtual machine.
+If no limit is specified for these values,
+then the representation must either be inefficient or vulnerable to overflow.
+The CPython virtual machine represents values like line numbers,
+stack offsets and instruction offsets by 32 bit values. This is inefficient, and potentially unsafe.
+
+It is inefficient as actual values rarely need more than a dozen or so bits to represent them.
+
+It is unsafe as malicious or poorly generated code could cause values to exceed 2\ :sup:`32`.
+
+For example, line numbers are represented by 32 bit values internally. 
+This is inefficient, given that modules almost never exceed a few thousand lines.
+Despite being inefficent, is is still vulnerable to overflow as
+it is easy for an attacker to created a module with billions of newline characters.
+
+Memory access is usually a limiting factor in the performance of modern CPUs.
+Better packing of data structures enhances locality and reduces memory bandwith,
+at a modest increase in ALU usage (for shifting and masking).
+Being able to safely store important values in 20 bits would allow memory savings
+in several data structures including, but not limited to:
+
+* Frame objects
+* Object headers
+* Code objects
+
+There is also the potential for a more efficient instruction format, speeding up interpreter dispatch.
```

Comment:
This is the ARM/IBM mistake all over again.

Reply:
Could you elaborate? Please bear in mind that this is software dispatching and decoding, not hardware.